P0009R12 mdspan changes

nliber opened the pull request:
These are the changes to the mdspan implementation so that it matches [P0009R12](https://wg21.link/P0009R12).

Note:  This is still missing changes to `stride()`.

nliber committed:
Changed ptrdiff_t to size_t and index_type to size_type

WIP: Compiles cleanly under clang12, but has warnings under gcc10

nliber committed:
Changed ptrdiff_t to size_t and index_type to size_type

WIP: Compiles cleanly under clang12, but has warnings under gcc10

nliber committed:
Fix for gcc narrowing warning when passing signed ints
to extents(Integral...) constructor

nliber committed:
Fix for gcc narrowing warning when passing signed ints
to extents(Integral...) constructor

nliber committed:
Renamed IndexType to SizeType or SizeTypes (depending if it is a single type or a parameter pack)

nliber committed:
Renamed IndexType to SizeType or SizeTypes (depending if it is a single type or a parameter pack)

nliber committed:
Made operator== and operator!= hidden friends

nliber committed:
Made operator== and operator!= hidden friends

nliber committed:
Added a compilation test for trivially copyable, which is
just a copy of ctest_standard_layout with is_standard_layout
instantiations replaces with is_trivially_copyable instantiations.

nliber committed:
Added a compilation test for trivially copyable, which is
just a copy of ctest_standard_layout with is_standard_layout
instantiations replaces with is_trivially_copyable instantiations.

nliber committed:
Renamed subspan to submdspan, as this only applies to mdspan

nliber committed:
Renamed subspan to submdspan, as this only applies to mdspan

nliber committed:
Renamed all_type to full_extent_t and all to full_extent

nliber committed:
Renamed all_type to full_extent_t and all to full_extent

nliber committed:
Renamed accessor_basic to default_accessor

nliber committed:
Renamed accessor_basic to default_accessor

nliber committed:
Removed accessor policy decay(p) member function as it was an artifact from an earlier version of this proposal when basic_mdspan had a span() member function that returned a std::span

nliber committed:
Removed accessor policy decay(p) member function as it was an artifact from an earlier version of this proposal when basic_mdspan had a span() member function that returned a std::span

nliber committed:
Added converting constructor for default_accessor (when the pointers to elements are convertible) for things like default_accessor<double> to default_accessor<const double>, mdspan<double, dynamic_extent> to mdspan<double const, dynamic_extent>, etc.

nliber committed:
Added converting constructor for default_accessor (when the pointers to elements are convertible) for things like default_accessor<double> to default_accessor<const double>, mdspan<double, dynamic_extent> to mdspan<double const, dynamic_extent>, etc.

nliber committed:
Changed std::experimental::all to std::experimental::full_extent in
benchmarks

nliber committed:
Changed std::experimental::all to std::experimental::full_extent in
benchmarks

nliber committed:
Changed ptrdiff_t to size_t across the board in benchmarks

nliber committed:
Changed ptrdiff_t to size_t across the board in benchmarks

nliber committed:
Fixed default_accessor breakage from rebase

nliber committed:
Merge branch 'P0009R12' of github.com:nliber/mdspan into P0009R12

dalg24 commented on include/experimental/__p0009_bits/default_accessor.hpp:
```diff
@@ -45,47 +45,43 @@
     class OtherElementType,
     /* requires */ (
-      _MDSPAN_TRAIT(is_convertible, typename accessor_basic<OtherElementType>::pointer, pointer)
+      _MDSPAN_TRAIT(is_convertible, OtherElementType, ElementType)
```
Are you sure about that change?

dalg24 commented on include/experimental/__p0009_bits/layout_left.hpp:
```diff
@@ -125,14 +125,14 @@ using idx_seq = make_index_sequence<sizeof...(Exts)>;
   // TODO @proposal-bug these (and other analogous operators) should be non-member functions
```
Drop the TODO?

nliber committed:
Fixed the constraint on the
default_accessor(default_accessor<OtherElementType>) constructor.

Removed a TODO in layout_left to make
comparison operators non-members, as it is now done.
